Added index to config, and ability to exclude/include top level dirs …

adelavega opened the pull request:
- To address #17 , include and exclude regex for top levels dir can be specified in "index" key of config.json.
- Added tests to cover folders that should be included/excluded, and all are passing

tyarkoni commented on grabbit/core.py:
```diff
@@ -185,9 +188,19 @@ def index(self):
+            #if they match exclude regex
+            if root == self.root:
+                if self.index_regex:
+                    print(root)
```
You forgot this print statement

tyarkoni replied:
Should be `if self.index_regex is not None`, I think

tyarkoni commented on grabbit/core.py:
```diff
@@ -185,9 +188,19 @@ def index(self):
+            # Exclude/include top-level directories from further search
+            #if they match exclude regex
+            if root == self.root:
+                if self.index_regex:
```
I don't see any reason why this should apply only to directories... probably all files should be screened against include and exclude. I'd probably move this into `_is_valid` (probably after the PR for validation is merged, to prevent conflicts). I.e., after each file is checked for validity in the base BIDS spec, we can then inspect it against the regexes (if needed) to see if it passes.

adelavega replied:
So do you mean the `_validate_file` function? So if you want to only exclude a directory you can a regex with a "/"

tyarkoni commented on grabbit/core.py:
```diff
@@ -185,9 +188,19 @@ def index(self):
+                    print(root)
+                    if 'exclude' in self.index_regex:
+                        for exclude_regex in self.index_regex['exclude']:
+                            directories[:] = [d for d in directories \
```
Make sure to `break` once a match is found; no need to waste cycles looping over all possible candidates once we know the candidate is to be excluded.

adelavega replied:
Once directories are excluded, the children are no longer looped over the way its written.

tyarkoni commented on grabbit/tests/specs/test.json:
```diff
@@ -1,4 +1,8 @@
 {
+  "index" : {
+    "exclude" : ["derivatives"],
```
Example feels a bit weird because there's no reason to check both include and exclude... if you define `include`, that implies that everything not in `include` is to be excluded, and vice versa for `exclude`. So we probably need at least one additional .json spec, so we can test both options. Also, maybe add more than one element to the list, to make sure it behaves properly in that case.

adelavega replied:
Yeah I agree.

adelavega commented:
Okay changes made! I updated the `_validate_file` method to check against include and exclude regexes, and also check directory names against those regexes to exclude them from further search. Also updated tests.

My only comment is that as implemented I don't find the "include" regex very useful. My original intention was to explicitly include BIDS folders such as "sub-*" and in the future maybe others. It's a lot more difficult now because the only way to differentiate folders from files is by extensions (e.g. `derivatives` is a folder but `derivatives.ext` is a file) and because regex is applied to all folders recursively. So in order to only include "sub-*" folders I also have to include sub-folders such as `func` `anat` in the regex for them to be included.

tyarkoni commented:
Great! Re; your last point, I'm not sure I follow... if the user wants to specify a directory rather than a file, they can end the regex with '$', no?

Also, keep in mind that this is a package that could in theory be used outside of pybids. There are lots of potential use cases where, e.g., you might have 10 sub-folders in a directory, but you only want to include 2 of them in the mapping. This just doesn't happen to be the BIDS use case.

adelavega commented:
Yes, but again that is assuming you differentiate files from folders with their name. Technically `derivatives` is a valid file name, so I don't see how you would differentiate that from a folder called `derivatives`. They are the same strings, no?

I agree with your second point, though.

adelavega commented:
@tyarkoni: I made the following changes, should be ready to go:

- added `_check_inclusions` and called on dirs and files 
- added an empty `_validate_dir` function that can overridden
- enforced that only include or exclude can specified in index + test
- option to override config at runtime by passing dictionary config + test

adelavega commented:
Also, after this merge can you do a release so we can merge my pybids PR?

tyarkoni commented on grabbit/core.py:
```diff
@@ -155,43 +153,81 @@ def __init__(self, path, config=None, dynamic_getters=False,
+
+        return True
+
+    def _validate_dir(self, d):
```
I'm thinking we should make this `return self._validate_file(d)` by default. In most cases, people will want to treat directories like files, so let's make them only have to override `_validate_dir` if they need *different* behavior for the two.

tyarkoni commented on grabbit/core.py:
```diff
@@ -155,43 +153,81 @@ def __init__(self, path, config=None, dynamic_getters=False,
             for f in filenames:
                 f = File(join(root, f))
-                if not self._validate_file(f):
+                if not (self._check_inclusions(f) or self._validate_file(f)):
```
Should be `and` instead of `or`, no? (Or alternatively, drop the parentheses.) We want to skip file if either condition fails.

tyarkoni commented on grabbit/tests/test_core.py:
```diff
@@ -174,6 +185,11 @@ def test_get_nearest(self, layout):
         assert nearest[0].subject == '01'
 
-    def test_exclude_regex(self, layout):
+    def test_index_regex(self, layout, layout2, layout_override):
```
We should also ensure that if you pass both include and exclude, it raises an exception.

adelavega commented on grabbit/core.py:
```diff
@@ -155,43 +153,81 @@ def __init__(self, path, config=None, dynamic_getters=False,
+                return False
+
+        # If file matches any excldue regex, then false
+        for regex in self.inclusion_regex.get('exclude', []):
```
Well since exclude and include are mutually exclusive, get should just return empty list so it won't actually check anything, but sure stylistically that is arguably better.

adelavega commented on grabbit/core.py:
```diff
@@ -155,43 +153,81 @@ def __init__(self, path, config=None, dynamic_getters=False,
 
         if config is not None:
-            self._load_config(config)
+            self._load_config(config, **kwargs)
```
But if you let the layout load the default config, the file indexing would have already happened, no?

adelavega commented:
@tyarkoni done

tyarkoni commented:
Cool. Did/can you check to make sure that the grabbids tests also pass when using this version of grabbit?

adelavega commented:
Yeah

adelavega commented:
Actually caught a minor bug that was causing master to fail, but now tests pass on both master and my PR on pybids.

tyarkoni commented:
Thanks!

tyarkoni commented:
Re: merging on pybids, let's wait on the other open PR (#26); I'd rather not have two minor releases very close to each other.

adelavega commented:
Makes sense
